src/redraw.mm

```diff
@@ -70,123 +70,65 @@ - (void) doAction:(RedrawCode::Enum)code withItem:(const object &)item_o
+        if (mReverseVideo) {
+            [textAttrs setObject:bg forKey:NSForegroundColorAttributeName];
+            bg = fg;
+        }
 
-        static std::vector<int> lens;
-        lens.clear();
+        // Draw background separately [width = item_sz]
+        NSRect bgrect = [self
+            viewRectFromCellRect:CGRectMake(mCursorPos.x, mCursorPos.y, item_sz - 1, 1)];
+        [bg set];
+        NSRectFill(bgrect);
 
-        static std::string run;
-        run.clear();
+        // Init string
+        NSString *nsrun = @"\u202d";
 
-        int len = 0;
-        for (int i=1; i<item_sz; i++) {
+        for (int i = 1; i < item_sz; i++) {
             const object &arglist = item_o.via.array.ptr[i];
 
             assert(arglist.via.array.size == 1);
-            const object &char_o = arglist.via.array.ptr[0];
-            std::string char_s = char_o.convert();
+            const std::string char_s = arglist.via.array.ptr[0].convert();
 
-            /* Vim gives us empty strings to pad out double-width characters.
-               Don't draw them as spaces (which would erase half the char); just
-               increase len of previous run. */
+            // Do nothing if last char was double width
             if (char_s.size() == 0) {
-                len += 1;
-                runs.push_back(run);
-                lens.push_back(len);
-                run = char_s;
-                len = 0;
-            }
-            else {
-                runs.push_back(run);
-                lens.push_back(len);
-                run = char_s;
-                len = 1;
+              continue;
             }
-        }
-
-        if (len) {
-            runs.push_back(run);
-            lens.push_back(len);
-        }
 
-        /* Swap FG & BG colours if necessary */
-        NSMutableDictionary *textAttrs;
-        if (mReverseVideo) {
-            textAttrs = [[[NSMutableDictionary alloc] init] autorelease];
-            [textAttrs setDictionary:mTextAttrs];
-            NSColor *fg = [textAttrs objectForKey:NSForegroundColorAttributeName];
-            NSColor *bg = [textAttrs objectForKey:NSBackgroundColorAttributeName];
-            [textAttrs setObject:fg forKey:NSBackgroundColorAttributeName];
-            [textAttrs setObject:bg forKey:NSForegroundColorAttributeName];
-        }
-        else {
-            textAttrs = mTextAttrs;
+            nsrun = [nsrun stringByAppendingString:
+              [NSString stringWithUTF8String:char_s.c_str()]];
         }
 
-        for (int i=0; i<runs.size(); i++) {
-            const std::string &run = runs[i];
-            int sz = lens[i];
+        // Force left-to-right rendering
+        nsrun = [nsrun stringByAppendingString:@"\u202c"];
 
-            if (sz == 0)
-                continue;
-
-            NSString *nsrun = [NSString stringWithUTF8String:run.c_str()];
+        CGContextSaveGState(mCanvasContext);
+        // TODO: Remove when font height formula is corrected (updateCharSize)
```

Comment:
Are you sure there is a problem with updateCharSize? Do you know what the problem is?